.doom.d/config.org
#+TITLE: Dot-Emacs Configuration
#+AUTHOR: Victor Vialard
#+STARTUP: fold

#+begin_src elisp
(mapc 'load (file-expand-wildcards "~/.doom.d/modules/*.el"))
(use-package exec-path-from-shell
  :ensure t
  :config
  (when (memq window-system '(mac ns x))
    (exec-path-from-shell-initialize)))
#+end_src

* General
** Personal Informations

#+begin_src elisp
(setq user-full-name "Victor Vialard"
      user-mail-address "vvialard@hotmail.fr")
(setq auth-sources '("~/.authinfo"))
#+end_src

** Constants

#+begin_src elisp
(if (eq system-type 'darwin)
    (setq emacs-cloud-root "/Users/sheepsody/Library/Mobile Documents/com~apple~CloudDocs/Emacs/")
  (setq emacs-cloud-root "/data/"))

(let ((default-directory emacs-cloud-root))
  (setq org-directory (expand-file-name "emacs/Org/")
        org-roam-directory (expand-file-name "emacs/Roam/")
        elfeed-db-directory (expand-file-name "emacs/.elfeed/")
        rmh-elfeed-org-files (list (expand-file-name "emacs/Org/elfeed/elfeed.org"))))
#+end_src

** Theme

#+begin_src elisp
;; Doom's theme
(load-theme 'catppuccin :no-confirm)
(setq doom-theme 'catppuccin)
(setq catppuccin-flavor 'latte) ;; or 'frappe, 'latte, 'macchiato, or 'mocha
(setq display-line-numbers 'relative)
(catppuccin-reload)
#+end_src

*** Switch dark/light version

#+begin_src elisp :tangle no
;; Synchronize between light and dark version
(defun synchronize-theme ()
  (let* ((light-theme 'latte)
         (dark-theme 'macchiato)
         (start-time-light-theme 6)
         (end-time-light-theme 18)
         (hour (string-to-number (substring (current-time-string) 11 13)))
         (next-theme (if (member hour (number-sequence start-time-light-theme end-time-light-theme))
                         light-theme dark-theme)))
    (when (not (equal catppuccin-flavor next-theme))
      (setq catppuccin-flavor next-theme)
      (catppuccin-reload))))
;; Disabled
(run-with-timer 0 900 'synchronize-theme)
#+end_src

*** Random splash image

#+begin_src elisp
;; Random
(setq random-splash-image-dir (concat "~/.doom.d/splash-images"))
(random-splash-image-set)
#+end_src

** Autocompletion

#+begin_src elisp
(after! company
  (setq company-idle-delay 0.2
        company-minimum-prefix-length 1)
  (setq company-show-numbers t)
  (add-hook 'evil-normal-state-entry-hook #'company-abort )) ;; make aborting less annoying.

(setq doom-modeline-enable-word-count nil)
#+end_src

** Bindings

#+begin_src elisp
;; Insert quotes for citations
(defun my/insert-quote ()
  (interactive)
  (insert "#+begin_quote\n«  »\n#+end_quote"))

(map! :leader
      :desc "Insert quote"
      "i q" #'my/insert-quote)

(map! :leader
      :desc "Structure Template"
      "i s" #'org-insert-structure-template)

(map! :leader
      :desc "Insert Reference"
      "i c" #'org-ref-helm-insert-cite-link)

(map! :leader
      :desc "Open Calendar"
      "o C" #'=calendar)

(map! :leader
      :desc "List bookmarks"
      "b L" #'list-bookmarks)
#+end_src

** Git

#+begin_src elisp
(global-git-gutter-mode +1)
#+end_src

** Dap Mode

#+begin_src elisp
;; Source: [[https://docs.doomemacs.org/latest/modules/tools/debugger/#keybindings,code-1--1][Doom emacs]]
(map! :map dap-mode-map
      :leader
      :prefix ("d" . "dap")
      ;; basics
      :desc "dap next"          "n" #'dap-next
      :desc "dap step in"       "i" #'dap-step-in
      :desc "dap step out"      "o" #'dap-step-out
      :desc "dap continue"      "c" #'dap-continue
      :desc "dap hydra"         "h" #'dap-hydra
      :desc "dap debug restart" "r" #'dap-debug-restart
      :desc "dap debug"         "s" #'dap-debug

      ;; debug
      :prefix ("dd" . "Debug")
      :desc "dap debug recent"  "r" #'dap-debug-recent
      :desc "dap debug last"    "l" #'dap-debug-last

      ;; eval
      :prefix ("de" . "Eval")
      :desc "eval"                "e" #'dap-eval
      :desc "eval region"         "r" #'dap-eval-region
      :desc "eval thing at point" "s" #'dap-eval-thing-at-point
      :desc "add expression"      "a" #'dap-ui-expressions-add
      :desc "remove expression"   "d" #'dap-ui-expressions-remove

      :prefix ("db" . "Breakpoint")
      :desc "dap breakpoint toggle"      "b" #'dap-breakpoint-toggle
      :desc "dap breakpoint condition"   "c" #'dap-breakpoint-condition
      :desc "dap breakpoint hit count"   "h" #'dap-breakpoint-hit-condition
      :desc "dap breakpoint log message" "l" #'dap-breakpoint-log-message)
#+end_src

** Path

#+begin_src elisp
(setenv "PATH" (concat ".:/opt/homebrew/bin/" (getenv "PATH")))
(setq exec-path (append exec-path '(".:/opt/homebrew/bin/")))
#+end_src

* Org-Mode
** Org

#+begin_src elisp
;; Org Directory
(setq org-image-actual-width '(600))
(custom-set-variables
 '(org-agenda-files (list
                     (concat org-directory "gtd/inbox.org")
                     (concat org-directory "gtd/habits.org")
                     (concat org-directory "gtd/calendar.org")
                     (concat org-directory "gtd/gtd.org"))))

(after! org
  :ensure t
  :config
  (setq org-id-extra-files (directory-files-recursively org-roam-directory "\.org$"))
  (setq org-archive-location (concat org-directory "archive.org::* FILE: %s"))
  (setq org-default-notes-file (concat org-directory "gtd/inbox.org"))
  (setq org-columns-default-format "%50ITEM(Task) %TODO %2PRIORITY %10Effort(Effort){:} %10CLOCKSUM %TAGS")

  ;; Save org buggers periodically
  (add-hook 'auto-save-hook 'org-save-all-org-buffers)

  ;; Inherit deadlines !
  (defun org-entry-properties-inherit-deadline (orig-fun &optional pom which)
    "Call ORIG-FUN with POM, but if WHICH is `DEADLINE' do it recursively."
    (if (string= which "SCHEDULED")
        (org-with-point-at pom
          (let (value)
            (while (not (or (setq value (funcall orig-fun (point) which))
                            (not (org-up-heading-safe)))))
            value)
          (funcall orig-fun pom which))))
  (advice-add 'org-entry-properties :around #'org-entry-properties-inherit-deadline)

  ;; Reduce size of latex fragments
  (plist-put org-format-latex-options :scale 3)

  ;; Open pdf in emacs
  (add-to-list 'org-file-apps '("\\.pdf\\'" . org-pdfview-open))

  ;; Capture templates
  (setq org-capture-templates
        `(
          ("t" "Todo [inbox]" entry
           (file ,(concat org-directory "gtd/inbox.org"))
           "* TODO %i%?")
          ("o" "One-Off [gtd]" entry
           (file+headline ,(concat org-directory "gtd/gtd.org") "One-Off")
           "* %^{Event status?|TODO|NEXT|WAITING|DONE } %i%?")
          ("c" "Org-protocol capture" entry
           (file ,(concat org-directory "gtd/inbox.org"))
           "* %a\n/Entered on:/ %U\n"
           :immediate-finish t)))

  ;; Keywords
  (setq org-todo-keywords
        '((sequence "TODO(t)" "NEXT(n)" "MEETING(m)" "WAITING(w)" "HOLD(h)" "|" "DONE(d)" "CANCELLED(c)")))
  (setq org-todo-keyword-faces
        '(("TODO"      :foreground "#04a5e5") ;; Sky
          ("NEXT"      :foreground "#fe640b") ;; Peach
          ("MEETING"   :foreground "#ea76cb") ;; Pink
          ("WAITING"   :foreground "#7287fd") ;; Lavender
          ("HOLD"      :foreground "#5c5f77") ;; Subtext
          ("DONE"      :foreground "#40a02b") ;; Green
          ("CANCELLED" :foreground "#5c5f77") ;; Subtext
          )))

(setq org-fancy-priorities-list '((?A . "❗")
                                  (?B . "⚡")
                                  (?C . "☕")))

(defun my-org-summarize-task-status ()
  "Count number of tasks by status.
      Probably should make this a dblock someday."
  (interactive)
  (let (result)
    (org-map-entries
     (lambda ()
       (let ((todo (elt (org-heading-components) 2)))
         (if todo
             (if (assoc todo result)
                 (setcdr (assoc todo result)
                         (1+ (cdr (assoc todo result))))
               (setq result (cons (cons todo 1) result)))))))
    (message "%s" (mapconcat (lambda (x) (format "%s: %d" (car x) (cdr x)))
                             result " | "))))
;; Stylesheet for HTML exports
(setq org-html-head "<link rel=\"stylesheet\" href=\"https://rawgithub.com/edwardtufte/tufte-css/gh-pages/tufte.css\" type=\"text/css\" />")

;; PlantUML
(setq plantuml-default-exec-mode 'jar)

;; Org-Protocol
(server-start)
(require 'org-protocol)
#+end_src

#+begin_src elisp
;; Source : https://macowners.club/posts/org-capture-from-everywhere-macos/
;; Command : /opt/homebrew/bin/emacsclient -ne "(timu-func-make-capture-frame)"
(defun timu-func-make-capture-frame ()
  "Create a new frame and run `org-capture'."
  (interactive)
  (make-frame '((name . "capture")
                (top . 300)
                (left . 700)
                (width . 80)
                (height . 25)))
  (select-frame-by-name "capture")
  (delete-other-windows)
  (noflet ((switch-to-buffer-other-window (buf) (switch-to-buffer buf)))
          (org-capture)))

(defadvice org-capture-finalize
    (after delete-capture-frame activate)
  "Advise capture-finalize to close the frame."
  (if (equal "capture" (frame-parameter nil 'name))
      (delete-frame)))

(defadvice org-capture-destroy
    (after delete-capture-frame activate)
  "Advise capture-destroy to close the frame."
  (if (equal "capture" (frame-parameter nil 'name))
      (delete-frame)))
#+end_src

#+begin_src bash :tangle no
# To toggle latex fragments, I had to run the following commands after installing MacTex
sudo ln -s /Library/TeX/texbin/latex /usr/local/bin/
sudo ln -s /Library/TeX/texbin/dvipng /usr/local/bin/
#+end_src

** Org-Pomodoro

#+begin_src elisp
(use-package org-pomodoro
  :after org
  :ensure t
  :commands (org-pomodoro)
  :config
  (setq
   org-pomodoro-length 25
   org-pomodoro-short-break-length 5
   org-pomodoro-long-break-length 20
   org-pomodoro-play-sounds nil))

(defun my/emacs-query-pomodoro ()
  "Return the remaining pomodoro time"
  (if (org-pomodoro-active-p)
      (cl-case org-pomodoro-state
        (:pomodoro
         (format "Pomo: %d min \n---\n%s" (/ (org-pomodoro-remaining-seconds) 60) org-clock-heading))
        (:short-break
         (format "Break: %d min" (/ (org-pomodoro-remaining-seconds) 60)))
        (:long-break
         (format "Break: %d min" (/ (org-pomodoro-remaining-seconds) 60)))
        (:overtime
         (format "Over! %d min" (/ (org-pomodoro-remaining-seconds) 60))))
    "No pomo"))
#+end_src

#+begin_src elisp
;; Needs terminal-notifier (brew install terminal-notifier)
(defun notify-osx (title msg)
  (let ((notifier-path (executable-find "terminal-notifier")))
       (start-process
           "OSX Alert"
           "*OSX Alert*" ; use `nil` to not capture output; this captures output in background
           notifier-path
           "-message" msg
           "-title" title
           "-sender" "com.apple.automator.EmacsClient")))

;; org-pomodoro mode hooks
(add-hook 'org-pomodoro-finished-hook
          (lambda ()
          (notify-osx "Pomodoro completed!" "Time for a break.")))

(add-hook 'org-pomodoro-break-finished-hook
          (lambda ()
          (notify-osx "Pomodoro Short Break Finished" "Ready for Another?")))

(add-hook 'org-pomodoro-long-break-finished-hook
          (lambda ()
            (notify-osx "Pomodoro Long Break Finished" "Ready for Another?")))

(add-hook 'org-pomodoro-killed-hook
          (lambda ()
          (notify-osx "Pomodoro Killed" "One does not simply kill a pomodoro!")))
#+end_src

#+RESULTS:
| lambda | nil | (notify-osx Pomodoro Killed One does not simply kill a pomodoro!) |

** Org-Clock

#+begin_src elisp
;; Display time & Org Clock config
(display-time-mode 1)
(setq org-global-properties
      '(("Effort_ALL" .
         "0:15 0:30 0:45 1:00 2:00 3:00 4:00 5:00 6:00 0:00"))
      org-clock-out-when-done t
      org-clock-out-remove-zero-time-clocks t)

;; Org-Clock export to csv
;; https://github.com/atheriel/org-clock-csv
(setq org-clock-csv-header "task,parents,category,start,end,duration,effort,ishabit,tags,itags"
      org-clock-csv-row-fmt (lambda (plist)
                              (mapconcat #'identity
                                         (list
                                          (org-clock-csv--escape (plist-get plist ':task))
                                          (org-clock-csv--escape (s-join org-clock-csv-headline-separator (plist-get plist ':parents)))
                                          (org-clock-csv--escape (plist-get plist ':category))
                                          (plist-get plist ':start)
                                          (plist-get plist ':end)
                                          (plist-get plist ':duration)
                                          (plist-get plist ':effort)
                                          (plist-get plist ':ishabit)
                                          (plist-get plist ':tags)
                                          (org-clock-csv--escape
                                           (org-clock-csv--read-property plist "ARCHIVE_ITAGS")))
                                         ",")))
#+end_src

** Org-Agenda

#+begin_src elisp
;; Org Super Agenda & Custom views
(use-package! org-super-agenda
  :commands (org-super-agenda-mode))

(add-hook 'org-super-agenda-mode-hook (lambda () (setq display-line-numbers 'relative)))
(add-hook 'org-agenda-mode-hook (lambda () (setq display-line-numbers 'relative)))

(after! org-agenda
  :init
  (setq org-agenda-skip-scheduled-if-done t
        org-agenda-skip-deadline-if-done t
        org-agenda-include-deadlines t
        org-agenda-block-separator nil
        org-agenda-compact-blocks t
        org-agenda-start-day nil ;; i.e. today
        org-agenda-span 1
        org-agenda-start-on-weekday nil)

  (setq org-agenda-custom-commands
        '(("z" "Custom GTD view"
           (
            (agenda "" ((org-agenda-overriding-header "")
                        (org-super-agenda-groups
                         '(
                           (:discard (:habit t :tag "habit"))
                           (:name "Today"
                            :time-grid t
                            :scheduled today
                            :order 1)))))
            (alltodo "" ((org-agenda-overriding-header "")
                         (org-agenda-skip-function '(org-agenda-skip-entry-if 'deadline))
                         (org-super-agenda-groups
                          '(
                            (:discard (:scheduled future))
                            (:name "Habits"
                             :habit t
                             :tag "habit"
                             :order 11)
                            (:discard (:habit t))
                            (:name "Waiting"
                             :todo "WAITING"
                             :order 25)
                            (:name "Immediatly"
                             :and (:priority "A" :todo "NEXT")
                             :order 2)
                            (:name "Next to do"
                             :todo "NEXT"
                             :order 3)
                            (:name "Due Today"
                             :deadline today
                             :order 4)
                            (:name "Important"
                             :tag "Important"
                             :priority "A"
                             :order 5)
                            (:name "Overdue"
                             :deadline past
                             :order 6)
                            (:name "Org Journal"
                             :file-path "Journal/"
                             :order 8)
                            (:name "Due Soon"
                             :deadline future
                             :order 9)
                            (:name "Less important"
                             :priority "B"
                             :order 10)
                            (:name "Trivial"
                             :tag "trivial"
                             :priority<= "C"
                             :order 30)
                            (:name "Courses"
                             :tag "course"
                             :order 12)
                            (:name "To read"
                             :tag "read"
                             :order 13)
                            (:name "Shopping"
                             :tag "shop"
                             :order 14)
                            (:name "To do"
                             :todo "TODO"
                             :order 20)
                            ))))))))
  :config
  (org-super-agenda-mode))
#+end_src

** Org-Journal

#+begin_src elisp
(use-package! org-journal
  :after org
  :ensure t
  :config
  (setq org-journal-dir (concat org-directory "Journal/")
        org-journal-file-type 'weekly
        org-journal-date-prefix "* "
        org-journal-file-format "%Y-%W.org"
        org-journal-date-format "%A, %B %d %Y"
        org-journal-enable-agenda-integration t
        org-journal-file-header "#+TITLE: Weekly Journal\n#+SUBTITLE: Week %W, %B %Y\n#+STARTUP: folded\n\n")

  (setq org-journal-carryover-items
        (string-join
         '("TODO=\"TODO\""
           "TODO=\"NEXT\""
           "TODO=\"WAITING\""
           "TODO=\"HOLD\"")
         "|"))

  (defvar org-journal--date-location-scheduled-time nil)

  (defun org-journal-date-location (&optional scheduled-time)
    (let ((scheduled-time (or scheduled-time (org-read-date nil nil nil "Date:"))))
      (setq org-journal--date-location-scheduled-time scheduled-time)
      (org-journal-new-entry t (org-time-string-to-time scheduled-time))
      (unless (eq org-journal-file-type 'daily)
        (org-narrow-to-subtree))
      (goto-char (point-max))))

  (defun my/refile (file headline)
    (let ((pos (function org-journal-find-location)))
      (org-refile nil nil (list headline file nil pos))))

  (add-to-list 'org-capture-templates
               `("j" "Journal entry" plain (function org-journal-date-location)
                 "** TODO %?\n SCHEDULED: <%(princ org-journal--date-location-scheduled-time)>\n"
                 :jump-to-captured t))
  (add-to-list 'org-capture-templates
               `("d" "Daily journal" plain (function org-journal-find-location)
                 (file ,(expand-file-name (concat org-directory "methodology/daily.org")))
                 :jump-to-captured t :immediate-finish t))
  (add-to-list 'org-capture-templates
               `("w" "Weekly review" plain (function org-journal-find-location)
                 (file ,(expand-file-name (concat org-directory "methodology/weekly.org")))
                 :jump-to-captured t :immediate-finish t))

  ;; My custom bindings
  (map! :leader
        (:prefix ("n j" . "Journal")
         :desc "New journal entry" "j" #'org-journal-new-entry
         :desc "New journal entry" "o" #'org-journal-open-current-journal-file
         :desc "Search journal entry" "s" #'org-journal-search)))
#+end_src

** Org-Roam

#+begin_src elisp
;; Org Roam

(setq org-roam-notes-path org-roam-directory)

(defun my-copy-file (source)
  "Copy a file from SOURCE to DESTINATION."
  (interactive "fCopy file: \nFDestination: ")
  (copy-file source "~/test1.dot"))

(use-package! org-roam
  :ensure t
  :config
  (setq org-roam-graph-executable "dot"
        org-roam-graph-filetype "dot"

.doom.d/config.el
(setq user-full-name "Victor Vialard"
      user-mail-address "vvialard@hotmail.fr")

;; Doom's theme
(setq doom-theme 'doom-one)
(setq display-line-numbers-type 'relative)

;; Start a good morning...
(setq fancy-splash-image "~/.doom.d/totoro.png")

(after! company
  (setq company-idle-delay 0.2
        company-minimum-prefix-length 1)
  (setq company-show-numbers t)
  (add-hook 'evil-normal-state-entry-hook #'company-abort )) ;; make aborting less annoying.

(setq ispell-dictionary "en_GB")

;; Insert quotes for citations
(defun my/insert-quote ()
  (interactive)
  (insert "#+begin_quote\n«  »\n#+end_quote"))
(map! :leader
      :desc "Insert quote"
      "i q" #'my/insert-quote)

(map! :leader
      :desc "Structure Template"
      "i s" #'org-insert-structure-template)

;; Org Directory
(setq org-directory (expand-file-name "~/Dropbox/Org/"))
(setq org-image-actual-width '(600))

(after! org
  (setq org-default-notes-file (concat org-directory "inbox.org"))
  (setq org-archive-location (concat org-directory "archived.org::"))

  (setq org-agenda-files '("~/Dropbox/Org/inbox.org"
                           "~/Dropbox/Org/tickler.org"
                           "~/Dropbox/Org/gtd.org"))

  ;; Capture templates
  (setq org-capture-templates
        '(("t" "Todo [inbox]" entry
           (file+headline "~/Dropbox/Org/inbox.org" "Tasks")
           "* TODO %i%?")
          ("T" "Tickler" entry
           (file+headline "~/Dropbox/Org/tickler.org" "Tickler")
           "* %i%? \n %U")))

  ;; Keywords
  (setq org-todo-keywords
        '((sequence "TODO(t)" "NEXT(n)" "MEETING(m)" "WAITING(w)" "HOLD(h)" "|" "DONE(d)" "CANCELLED(c)")))
  (setq org-todo-keyword-faces
        '(("TODO"      :foreground "#46D9FF") ;; brightblue
          ("NEXT"      :foreground "#da8548") ;; brightred
          ("MEETING"   :foreground "#ECBE7B") ;; yellow
          ("WAITING"   :foreground "#c678dd") ;; brightmagenta
          ("HOLD"      :foreground "#a9a1e1") ;; magenta
          ("DONE"      :foreground "#98be65") ;; brightgreen
          ("CANCELLED" :foreground "#73797e") ;; gray
          )))

;; Display time & Org Clock config
(display-time-mode 1)
(setq org-global-properties
      '(("Effort_ALL" .
         "0:15 0:30 0:45 1:00 2:00 3:00 4:00 5:00 6:00 0:00")))

;; Org Super Agenda & Custom views
(use-package! org-super-agenda
  :commands (org-super-agenda-mode))

(after! org-agenda
  :init
  (setq org-agenda-skip-scheduled-if-done t
        org-agenda-skip-deadline-if-done t
        org-agenda-include-deadlines t
        org-agenda-block-separator nil
        org-agenda-compact-blocks t
        org-agenda-start-day nil ;; i.e. today
        org-agenda-span 1
        org-agenda-start-on-weekday nil)

  (setq org-agenda-custom-commands
        '(("z" "Custom GTD view"
           ((agenda "" ((org-agenda-overriding-header "")
                        (org-super-agenda-groups
                         '((:name "Today"
                            :discard (:tag ("daily" "weekly"))
                            :time-grid t
                            :date today
                            :order 1)))))
            (alltodo "" ((org-agenda-overriding-header "")
                         (org-super-agenda-groups
                          '(
                            (:name "Dailies"
                             :and (:tag "daily"
                                   :not (:scheduled future))
                             :order 9)
                            (:name "Weeklies"
                             :and (:tag "weekly"
                                   :not (:scheduled future))
                             :order 10)
                            (:discard (:tag ("daily" "weekly")))

                            (:name "Immediatly"
                             :and (:priority "A" :todo "NEXT")
                             :order 1)
                            (:name "Next to do"
                             :todo "NEXT"
                             :order 2)
                            (:name "Due Today"
                             :deadline today
                             :order 3)
                            (:name "Important"
                             :tag "Important"
                             :priority "A"
                             :order 4)
                            (:name "Overdue"
                             :deadline past
                             :order 5)
                            (:name "Due Soon"
                             :deadline future
                             :order 6)
                            (:name "Less important"
                             :priority "B"
                             :order 7)
                            (:name "Trivial"
                             :tag "trivial"
                             :priority<= "C"
                             :order 30)
                            (:name "Courses"
                             :tag "course"
                             :order 12)
                            (:name "To read"
                             :tag "read"
                             :order 13)
                            (:name "Shopping"
                             :tag "shop"
                             :order 14)
                            (:name "Courses"
                             :tag "course"
                             :order 15)
                            (:name "To do"
                             :todo "TODO"
                             :order 20)
                            (:name "Waiting"
                             :todo "WAITING"
                             :order 25)
                            ))))))))
  :config
  (org-super-agenda-mode))

(use-package! org-journal
  :ensure t
  :defer t
  :config
  (setq org-journal-dir "~/Dropbox/Journal/"
        org-journal-file-type 'weekly
        org-journal-date-prefix "* "
        org-journal-file-format "%Y-%m-%d.org"
        org-journal-date-format "%A, %B %d %Y"
        org-journal-file-header "#+TITLE: Weekly Journal\n#+SUBTITLE: Week %U, %B %Y\n#+STARTUP: folded\n\n")
  (defun org-journal-find-location ()
    (org-journal-new-entry t)
    (unless (eq org-journal-file-type 'daily)
      (org-narrow-to-subtree))
    (goto-char (point-max)))
  (add-to-list 'org-capture-templates
               `("j" "Journal entry" entry (function org-journal-find-location)
                 "** %(format-time-string org-journal-time-format)%^{Title}\n%i%?"
                 :jump-to-captured t :immediate-finish t))
  (add-to-list 'org-capture-templates
               `("w" "Weekly review" entry (function org-journal-find-location)
                 (file ,(expand-file-name "~/Dropbox/Org/weekly.org"))
                 :jump-to-captured t :immediate-finish t)))

;; Org Roam
;; Inspired from https://github.com/alexkehayias/emacs.d/

(setq org-roam-publish-path "~/GitHub/MySecondBrainWebsite")
(setq org-roam-notes-path "~/Dropbox/Roam")

(setq org-roam-directory "~/Dropbox/Roam/")
(use-package org-roam
  :config
  (setq org-roam-graph-executable "neato")
  (setq org-roam-graph-extra-config
        '(("overlap" . "false")))
  :init
  ;; These functions need to be in :init otherwise they will not be
  ;; callable in an emacs --batch context which for some reason
  ;; can't be found in autoloads if it's under :config
  (defun my/org-roam--extract-note-body (file)
    (with-temp-buffer
      (insert-file-contents file)
      (org-mode)
      (first (org-element-map (org-element-parse-buffer) 'paragraph
               (lambda (paragraph)
                 (let ((begin (plist-get (first (cdr paragraph)) :begin))
                       (end (plist-get (first (cdr paragraph)) :end)))
                   (buffer-substring begin end)))))))

  ;; Include backlinks in org exported notes not tagged as private or
  ;; draft
  (defun my/org-roam--backlinks-list (file)
    (if (org-roam--org-roam-file-p file)
        (--reduce-from
         (concat acc (format "- [[file:%s][%s]]\n"
                             (file-relative-name (car it) org-roam-directory)
                             (org-roam-db--get-title (car it))))
         ""
         (org-roam-db-query
          [:select :distinct [links:source]
           :from links
           :left :outer :join tags :on (= links:source tags:file)
           :where (and (= dest $s1)
                       (or (is tags:tags nil)
                           (and
                            (not-like tags:tags '%private%)
                            (not-like tags:tags '%draft%))))]
          file))
      ""))

  (defun file-path-to-md-file-name (path)
    (let ((file-name (first (last (split-string path "/")))))
      (concat (first (split-string file-name "\\.")) ".md")))

  ;; Fetches all org-roam files and exports to hugo markdown
  ;; files. Adds in necessary hugo properties
  ;; e.g. HUGO_BASE_DIR. Ignores notes tagged as private or draft
  (defun org-roam-to-hugo-md ()
    (interactive)
    ;; Make sure the author is set
    (setq user-full-name "Victor Vialard")

    (let ((files (mapcan
                  (lambda (x) x)
                  (org-roam-db-query
                   [:select [files:file]
                    :from files
                    :left :outer :join tags :on (= files:file tags:file)
                    :where (or (is tags:tags nil)
                               (and
                                (not-like tags:tags '%private%)
                                (not-like tags:tags '%draft%)))]))))
      (mapc
       (lambda (f)
         ;; Use temporary buffer to prevent a buffer being opened for
         ;; each note file.
         (with-temp-buffer
           (message "Working on: %s" f)
           (insert-file-contents f)

           (goto-char (point-min))
           ;; Add in hugo tags for export. This lets you write the
           ;; notes without littering HUGO_* tags everywhere
           ;; HACK:
           ;; org-export-output-file-name doesn't play nicely with
           ;; temp buffers since it attempts to get the file name from
           ;; the buffer. Instead we explicitely add the name of the
           ;; exported .md file otherwise you would get prompted for
           ;; the output file name on every note.
           (insert
            (format "#+HUGO_BASE_DIR: %s\n#+HUGO_SECTION: posts\n#+EXPORT_FILE_NAME: %s\n"
                    org-roam-publish-path
                    (file-path-to-md-file-name f)))

           ;; If this is a placeholder note (no content in the
           ;; body) then add default text. This makes it look ok when
           ;; showing note previews in the index and avoids a headline
           ;; followed by a headline in the note detail page.
           (if (eq (my/org-roam--extract-note-body f) nil)
               (progn
                 (goto-char (point-max))
                 (insert "\n/This note does not have a description yet./\n")))

           ;; Add in backlinks because
           ;; org-export-before-processing-hook won't be useful the
           ;; way we are using a temp buffer
           (let ((links (my/org-roam--backlinks-list f)))
             (unless (string= links "")
               (goto-char (point-max))
               (insert (concat "\n* Links to this note\n") links)))

           (org-hugo-export-to-md)))
       files)))

  (map! :leader
        :desc "Export Roam notes with Ox-Hugo"
        "n r h" #'org-roam-to-hugo-md)
  :config
  (setq org-roam-capture-templates
        '(("p" "private" plain #'org-roam--capture-get-point "%?"
           :file-name "%<%Y%m%d>-${slug}"
           :head "#+TITLE: ${title}\n#+date: %t\n#+hugo_lastmod: %t\n#+ROAM_TAGS: private\n\n"
           :unnarrowed t)
          ("d" "draft" plain #'org-roam--capture-get-point "%?"
           :file-name "%<%Y%m%d>-${slug}"
           :head "#+TITLE: ${title}\n#+date: %t\n#+hugo_lastmod: %t\n#+ROAM_TAGS: draft\n\n"
           :unnarrowed t))))

;; Deft Configuration
;; Used for quick browsing of plain text notes

(use-package deft
  :config
  (setq deft-extensions '("txt" "tex" "org"))
  (setq deft-directory "~/Dropbox/Roam")
  (setq deft-recursive t))

(use-package ox-hugo
  :config
  (setq org-hugo-date-format "%Y-%m-%d"))

;; Update timestamps automatically when saving !
(add-hook 'org-mode-hook (lambda ()
                           (setq-local time-stamp-active t
                                       time-stamp-line-limit 18
                                       time-stamp-start "^#\\+hugo_lastmod: [ \t]*"
                                       time-stamp-end "$"
                                       time-stamp-format "\[%Y-%m-%d %a %H:%M:%S\]")
                           (add-hook 'before-save-hook 'time-stamp nil 'local)))

(use-package citeproc-org
  :ensure t
  :after ox-hugo
  :config
  (citeproc-org-setup))


(use-package! org-ref
  :after org
  :config
  (setq org-ref-bibliography-notes "~/Dropbox/Roam/refs/"
        org-ref-default-bibliography '("~/Dropbox/Roam/references.bib")
        org-ref-pdf-directory "~/Dropbox/Roam/pdf/"))

(use-package! bibtex-completion
  :config
  (setq bibtex-completion-pdf-symbol "⌘"
        bibtex-completion-pdf-field "file"
        bibtex-completion-notes-symbol "✎"
        bibtex-completion-bibliography "~/Dropbox/Roam/references.bib"
        bibtex-completion-library-path "~/Dropbox/Roam/pdf/"
        bibtex-completion-notes-path "~/Dropbox/Roam/refs/"
        bibtex-completion-notes-template-multiple-files
        (concat
         "${title}\n"
         "#+hugo_lastmod: %t\n"
         "#+roam_key: cite:${=key=}\n"
         "#+ROAM_TAGS: private\n"
         ":PROPERTIES:\n"
         ":CUSTOM_ID: ${=key=}\n"
         ":NOTER_DOCUMENT: %(orb-process-file-field \"${=key=}\")\n"
         ":AUTHOR: ${author-abbrev}\n"
         ":JOURNAL: ${journaltitle}\n"
         ":DATE: ${date}\n"
         ":YEAR: ${year}\n"
         ":DOI: ${doi}\n"
         ":URL: ${url}\n"
         ":END:\n\n"
         )))

;; Tell org-ref to let helm-bibtex find notes for it
(setq org-ref-notes-function
      (lambda (thekey)
        (let ((bibtex-completion-bibliography (org-ref-find-bibliography)))
          (bibtex-completion-edit-notes
           (list (car (org-ref-get-bibtex-key-and-file thekey)))))))

;; Rust configutation

(setq lsp-rust-server 'rust-analyzer)

;; Js/Ts/CSS config

(use-package prettier-js
  :config
  (setq prettier-js-args '(
                           "--trailing-comma" "es5"
                           "--single-quote" "true"
                           "--print-width" "120"
                           "--tab-width" "4"
                           "--use-tabs" "false"
                           "--jsx-bracket-same-line" "false"
                           "--stylelint-integration" "true"
                           )))

(use-package js2-mode)

(use-package rjsx-mode
  :mode(("\\.js\\'" . rjsx-mode)
        ("\\.jsx\\'" . rjsx-mode))
  :init
  (add-hook 'rjsx-mode-hook 'prettier-js-mode)
  (add-hook 'rjsx-mode-hook 'tide-mode))

(use-package tide
  :mode(("\\.ts\\'" . typescript-mode))
  :init
  (add-hook 'typescript-mode-hook 'tide-mode)
  (add-hook 'typescript-mode-hook 'prettier-js-mode)
  :config
  (tide-setup)
  (flycheck-mode +1)
  (setq flycheck-check-syntax-automatically '(save-mode-enabled))
  (eldoc-mode +1)
  (tide-hl-identifier-mode +1)
  (company-mode +1))

;; Python Environnements

(use-package pyvenv
  :config
  (pyvenv-mode t)
  (setenv "WORKON_HOME" "~/.pyenv/versions")

  ;; Set correct Python interpreter
  (setq pyvenv-post-activate-hooks
        #'(lambda ()
            (call-interactively #'lsp-workspace-restart)
            (setq python-shell-interpreter (concat pyvenv-virtual-env "bin/python3"))))
  (setq pyvenv-post-deactivate-hooks
        (list (lambda ()
                (setq python-shell-interpreter "python3")))))

(setq org-html-head "<link rel=\"stylesheet\" href=\"https://rawgithub.com/edwardtufte/tufte-css/gh-pages/tufte.css\" type=\"text/css\" />")

;; C++ Config
(setq lsp-clients-clangd-args '("-j=3"
                                "--background-index"
                                "--clang-tidy"
                                "--completion-style=detailed"
                                "--header-insertion=never"))
(after! lsp-clangd (set-lsp-priority! 'clangd 2))

;; ElFeed RSS reader

(use-package! elfeed-org
  :ensure t
  :config
  (elfeed-org)
  (setq rmh-elfeed-org-files (list "~/Dropbox/Org/elfeed.org")))
(map! :leader
      (:prefix-map ("e" . "elfeed")
       :desc "ElFeed" "e" #'elfeed))
